crates/net/downloaders/src/bodies/concurrent.rs

```diff
@@ -1,43 +1,226 @@
+            let (_, header) = header_cursor.seek_exact(key)?.expect("database corrupted");
+            if !header.is_empty() {
+                non_empty_headers += 1;
+            }
+            headers.push(SealedHeader::new(header, key.hash()));
+            if non_empty_headers >= count || headers.len() >= self.stream_batch_size {
+                break
+            }
+            canonical_entry = canonical_cursor.next()?;
+        }
+
+        Ok(headers)
+    }
+
+    fn next_expected_block_number(&self) -> BlockNumber {
+        match self.latest_queued_block_number {
+            Some(num) => num + 1,
+            None => self.download_range.start,
+        }
+    }
+
+    /// Max requests to handle at the same time
+    ///
+    /// This depends on the number of active peers but will always be
+    /// [`min_concurrent_requests`..`max_concurrent_requests`]
+    #[inline]
+    fn concurrent_request_limit(&self) -> usize {
+        let num_peers = self.client.num_connected_peers();
+
+        // we try to keep more requests than available peers active so that there's always a
+        // followup request available for a peer
+        let dynamic_target = num_peers * CONCURRENCY_PEER_MULTIPLIER;
+        let max_dynamic = dynamic_target.max(*self.concurrent_requests_range.start());
+
+        // If only a few peers are connected we keep it low
+        if num_peers < *self.concurrent_requests_range.start() {
+            return max_dynamic
+        }
+
+        max_dynamic.min(*self.concurrent_requests_range.end())
+    }
+
+    fn is_terminated(&self) -> bool {
+        self.in_progress_queue
+            .last_requested_block_number
+            .map(|last| last + 1 == self.download_range.end)
+            .unwrap_or_default() &&
+            self.in_progress_queue.is_empty() &&
+            self.buffered_responses.is_empty() &&
+            self.queued_bodies.is_empty()
+    }
+
+    /// Queues bodies and sets the latest queued block number
+    fn queue_bodies(&mut self, bodies: Vec<BlockResponse>) {
+        self.latest_queued_block_number = Some(bodies.last().expect("is not empty").block_number());
+        self.queued_bodies.extend(bodies.into_iter());
+    }
+
+    /// Returns a response if it's first block number matches the next expected.
+    fn try_next_buffered(&mut self) -> Option<Vec<BlockResponse>> {
+        if let Some(next) = self.buffered_responses.peek() {
+            let expected = self.next_expected_block_number();
+            let next_block_rng = next.block_range();
+
+            if next_block_rng.contains(&expected) {
+                return self.buffered_responses.pop().map(|buffered| {
+                    buffered
+                        .0
+                        .into_iter()
+                        .skip_while(|b| b.block_number() < expected)
+                        .take_while(|b| self.download_range.contains(&b.block_number()))
+                        .collect()
```

Comment:
that's correct, but i am mindful of introducing footguns and this would be one of them. currently, until the stream is terminated, we always call `set_download_range` with the same range end value. if that were to change for any reason (some specifics of stage implementation), this code would incorrectly put extra bodies into the queue.

the dl implementation should not be concerned about how it's used.